File: java-workshop/part-1-powermock/src/test/java/service/PaymentServiceTest.java
package service;

import domain.PaymentResult;
import domain.PaymentStatus;
import infrastructure.PaymentGateway;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.MockedStatic;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

/**
 * PowerMock style: We need to mock the PaymentGateway static methods
 */
class PaymentServiceTest {

    private PaymentService paymentService;
    private MockedStatic<PaymentGateway> mockedGateway;

    @BeforeEach
    void setUp() {
        mockedGateway = mockStatic(PaymentGateway.class);
        paymentService = new PaymentService();
    }

    @AfterEach
    void tearDown() {
        mockedGateway.close();
    }

    @Test
    void processPayment_returnsSuccess_whenGatewaySucceeds() {
        // Given
        String orderId = "ORDER-123";
        String customerId = "CUST-456";
        double amount = 100.0;

        mockedGateway.when(() -> PaymentGateway.processPayment(customerId, amount))
            .thenReturn(new PaymentResult("", PaymentStatus.SUCCESS, "TXN-789"));

        // When
        PaymentResult result = paymentService.processPayment(orderId, customerId, amount);

        // Then
        assertEquals(PaymentStatus.SUCCESS, result.status());
        assertEquals(orderId, result.orderId());
        assertEquals("TXN-789", result.transactionId());

        mockedGateway.verify(() -> PaymentGateway.processPayment(customerId, amount), times(1));
    }

    @Test
    void processPayment_returnsFailure_forZeroAmount_withoutCallingGateway() {
        // Given
        String orderId = "ORDER-123";
        String customerId = "CUST-456";
        double amount = 0.0;

        // When
        PaymentResult result = paymentService.processPayment(orderId, customerId, amount);

        // Then
        assertEquals(PaymentStatus.FAILED, result.status());
        mockedGateway.verifyNoInteractions();
    }

    @Test
    void processPayment_returnsFailure_whenGatewayFails() {
        // Given
        String orderId = "ORDER-123";
        String customerId = "CUST-456";
        double amount = 100.0;

        mockedGateway.when(() -> PaymentGateway.processPayment(customerId, amount))
            .thenReturn(new PaymentResult("", PaymentStatus.FAILED, null));

        // When
        PaymentResult result = paymentService.processPayment(orderId, customerId, amount);

        // Then
        assertEquals(PaymentStatus.FAILED, result.status());
    }

    @Test
    void refundPayment_returnsTrue_whenGatewayRefundSucceeds() {
        String transactionId = "TXN-789";

        mockedGateway.when(() -> PaymentGateway.refund(transactionId)).thenReturn(true);

        boolean result = paymentService.refundPayment(transactionId);

        assertTrue(result);
        mockedGateway.verify(() -> PaymentGateway.refund(transactionId), times(1));
    }

    @Test
    void refundPayment_returnsFalse_whenPaymentWasAlreadyRefunded() {
        String transactionId = "TXN-789";

        mockedGateway.when(() -> PaymentGateway.refund(transactionId)).thenReturn(false);

        boolean result = paymentService.refundPayment(transactionId);

        assertFalse(result);
        mockedGateway.verify(() -> PaymentGateway.refund(transactionId), times(1));
    }
}

File: kotlin-workshop/part-1-powermock/src/test/kotlin/service/PaymentServiceTest.kt
package service

import domain.PaymentResult
import domain.PaymentStatus
import infrastructure.PaymentGateway
import io.mockk.*
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals

/**
 * PowerMock style: We need to mock the PaymentGateway object to test PaymentService
 */
class PaymentServiceTest {
    
    private lateinit var paymentService: PaymentService
    
    @BeforeEach
    fun setUp() {
        mockkObject(PaymentGateway) // Mock the static-like object!
        paymentService = PaymentService()
    }
    
    @AfterEach
    fun tearDown() {
        unmockkObject(PaymentGateway)
    }
    
    @Test
    fun `processPayment returns success when gateway succeeds`() {
        // Given
        val orderId = "ORDER-123"
        val customerId = "CUST-456"
        val amount = 100.0
        
        every { 
            PaymentGateway.processPayment(customerId, amount) 
        } returns PaymentResult("", PaymentStatus.SUCCESS, "TXN-789")
        
        // When
        val result = paymentService.processPayment(orderId, customerId, amount)
        
        // Then
        assertEquals(PaymentStatus.SUCCESS, result.status)
        assertEquals(orderId, result.orderId)
        assertEquals("TXN-789", result.transactionId)
        
        verify(exactly = 1) { PaymentGateway.processPayment(customerId, amount) }
    }
    
    @Test
    fun `processPayment returns failure for zero amount without calling gateway`() {
        // Given
        val orderId = "ORDER-123"
        val customerId = "CUST-456"
        val amount = 0.0
        
        // When
        val result = paymentService.processPayment(orderId, customerId, amount)
        
        // Then
        assertEquals(PaymentStatus.FAILED, result.status)
        verify(exactly = 0) { PaymentGateway.processPayment(any(), any()) }
    }
    
    @Test
    fun `processPayment returns failure when gateway fails`() {
        // Given
        val orderId = "ORDER-123"
        val customerId = "CUST-456"
        val amount = 100.0
        
        every { 
            PaymentGateway.processPayment(customerId, amount) 
        } returns PaymentResult("", PaymentStatus.FAILED)
        
        // When
        val result = paymentService.processPayment(orderId, customerId, amount)
        
        // Then
        assertEquals(PaymentStatus.FAILED, result.status)
    }
}

